core/embed/bootloader/main.c

```diff
@@ -294,6 +299,79 @@ static void check_bootloader_version(void) {
 
 #endif
 
+void failed_jump_to_firmware(const image_header *hdr, vendor_header *vhdr_ptr) {
+  error_shutdown("INTERNAL ERROR", "(glitch)");
+}
+
+void real_jump_to_firmware(const image_header *hdr, vendor_header *vhdr_ptr) {
+  // we do not need deep copy, C doesn't have references until C20
+  vendor_header vhdr = *vhdr_ptr;
```

Comment:
what is the purpose of copying vhdr here? and what is the comment telling me? :)

Reply:
this actually causes a segfault when installing firmware in the emulator.... working on `vhdr_ptr` fixes it

Reply:
Original reason was so that better diffs view that code as moved unlike removed and added with differences, but in C you can't do reference, like `vendor_header&`. In the end the `clang-format` reformatted it anyway, so just shallow copy seemed fine. BTW clang-format is very "non-deterministic in what it likes", it just makes code look "made by machine ugly".

So then changing it to pointer without the copy would be better (compiler generates memcpy there on ARM). I'll have to check what could cause segfault with shallow copy, since all calls are either `const vendor_header*` and the pointers work with the same buffers (original struct is never used again, since it's no-return, which should have been added as attribute probably).

Still weird that the shallow copy looks fine, vhdr used in const * functions, and then it segfaults. Can't catch it with watchpoints, will need to build it with less optimization since I can't see some variables.

Reply:
Apparently `vhdr.origin` points completely outside of process mapped range the moment `firmware_jump_fn` is reached:

![image](https://github.com/trezor/trezor-firmware/assets/812173/f2aa3833-caf6-4848-9fb7-e2de65ce9aa9)

Changed here:

![image](https://github.com/trezor/trezor-firmware/assets/812173/e9ddd640-da6b-4f20-921e-d84a335586b3)

And then for some reason the watchpoint was lost and dunno when changed.

Reply:
not sure what is going on, but there is actually zero reason to pass anything at all. see d27d256a6d5e3183a28f49da7c4497581c71df0f, @TychoVrahe please double-check but I believe this is how it's supposed to work

Reply:
The problem in the before-fix version was that `read_vendor_header` read the vendor header into location pointed at by vhdr_ptr, but the rest of the evaluation was based on the `vhdr` variable. In case of firmware just being installed this was filled by garbage, since there wasn't any vendor header before. 

The fix makes sense to me.

Reply:
Removed the comment and added initialization https://github.com/trezor/trezor-firmware/pull/3122/commits/e52f4c26efefb9c4ae5480ddebe8a284a5b0257c